cloudmarker/stores/mongodbstore.py

```diff
@@ -3,51 +3,67 @@
+                 buffer_size=1000, **kwargs):
         """Create an instance of :class:`MongoDBStore` plugin.
 
-        Arguments:
-            host (str): MongoDB instance host.
-            port (int): MongoDB instance port.
-            username (str): MongoDB instance auth username.
-            password (str): MongoDB instance auth password.
-            db (str): MongoDB database name, default: ``gcp``
-            buffer_size (int): Max number of records to hold in memeory for
-                each record_type.
+        It will use the default port for mongodb 27017 if not specified.
+        The Authentication scheme will be negotiated by MongoDB and the client
+        for v4.0+ to SCRAM-SHA-1 or SCRAM-SHA-256 by default aftere
+        negotiation.
 
+        Arguments:
+            db (str): name of the database
+            username (str): username for the database
+            password (str): password for username to authenticate with the db
+            host (str): hostname for the DB server
+            port (int): port for mongoDB is listening, defaults to 27017
+            buffer_size (int): max buffer before flushing to db
+            kwargs (dict): Additional args
+                * models - List of classes for validator and enforcement
+                    requirements.
         """
-        # pylint: disable=too-many-instance-attributes
-
-        self._mongodb_host = host
-        self._mongodb_port = port
-        self._mongodb_username = username
-        self._mongodb_password = password
-        self._db = db
+        self._client = MongoClient(
+            host=host,
+            port=port,
+            username=username,
+            password=password
+        )
+
+        self._db = self._client[db]
+
+        # Create collections if they don't exist based on the document
+        # restrictions the users want to have
+        """The collection creation, where the models objects specified, are
+        done with the following way. The object specified must have the
+        following functions:
+            collection(): returns the name of the collection
+            validator(): returns a valid string with $jsonSchema that contains
+                all the validation properties as specified in the mongodb
+                docs.
+            enforce(): returns one of the strings ``error`` or ``warn``.
+                This will be used to define the error level for when the schema
+                validation fails. ``error`` will raise an error while ``warn``
+                will log that violation in the db server logs.
+        """
+        for model in kwargs.get('models', []):
+            if model.collection() not in self._db.list_collection_names():
+                self._db.create_collection(
+                    model.collection(),
+                    validator=model.validator(),
+                    validationAction=model.enforce()
+                )
```

Comment:
I think this validation is not required for the problem we are trying to solve. We want all data to be entered into our data stores regardless of how they look. As long as the data is valid JSON, we enter it into our data stores. The reason is that we do not have control on the data sources. Whatever data is presented to us by the cloud provider is the data we have to save and work with. Further, in a security software like this usually all data needs to be saved regardless of whether they adhere to a specific schema or not.

My opinion is that this change should not be merged into `master`. I would like to know if you or anyone else wants this to be merged. Since this is a community project and we are a developer community of four right now, I think we can take a decision based on what the community thinks.

By the way, this is one reason why separate concerns should be separate pull requests. Right now, there are two commits in this pull request addressing separate concerns and if we decide not to merge this, then the whole pull request will remain blocked until this change is removed. Therefore, I request you to make each pull request further smaller in scope so that each can be reviewed and merged independently. At the very least, this validation change should be moved out to a separate pull request so that it can be discussed separately.

Reply:
The reason I want this feature is to allow basic validation into fields, say some `id`, or some `name` field. This would create one safeguard if someone tries to push something incorrect or non-relevant into this collection as this is just a collection level check.